docs/proposals/thin-volumes.md

```diff
@@ -0,0 +1,175 @@
+
+## Motivation
+- Storage space can be used more effectively. More users can be accommodated for the same amount of storage space when compared to thick provisioning. This significantly reduces upfront hardware cost for the storage admins.
+
+- Faster clones and snapshots.
+
+### Goals
+- As a storage admin, I want to dynamically provision disk space to users on demand based on how much space they need at a given time, so that I can manage unused disk space efficiently.
+
+- As a storage user,  I want to be able to create a large PV, so that I don’t have to worry about extending the PV over and over again.
+
+- As a storage admin, I want to be alerted if the available size of Volume Group with thin logical volumes crosses a certian threshold, so that I can take precautionary measures like extending the volume group.
+
+## Proposal
+- Topolvm is **not** responsible for creating the thin-pools. Storage admin/operator has to create the thin-pools in advance while creating the volume groups.
+- Multiple device classes can be mapped to a single volume group having multiple thin pools. But each device class should map to a single thin pool within that volume group.
+- `lvmd.yaml` file should be updated to hold the mapping between device classes, volume group and thin pools.
+- For device classes mapped to thin pools, the node annotation `capacity.topolvm.cybozu.com/<device-class>` should be updated with respect to the thin pool size and overprovision ratio.
+- Metrics related to thin volume `data` and `metadata` parameters should be exported.
+
+
+
+## Design Details
+- A storage admin has to create the volume groups and the thin-pools on this volume group.
+- The storage admin should update the lvmd.yaml config file on the nodes to map device classes, volume groups and the corresponding thin-pools.
+- Sample `lvmd.yaml` config file:
+``` yaml=
+device-classes:
+  - name: ssd-thin
+    volume-group: myvg1
+    default: true
+    spare-gb: 10
+    type: thin
+    thin-pool:
+        name: pool0
+        overprovision-ratio: 50.0
+  - name: ssd
+    volume-group: myvg1
+    default: true
+    spare-gb: 10
+```
+- Above example shows two device classes; `ssd-thin` and `ssd`. Both use `myvg1` volume group.
+- `ssd-thin` has thin pool parameters.
+- Only thin volumes will be created when `ssd-thin` device class is selected by the user in the storageClass parameter `capacity.topolvm.cybozu.com/device-class`
+- Only thick volumes will be created when user has specified `ssd` in the storageClass.
+
+#### API Changes
+
+- The `DeviceClass` struct of the device-class-manager will hold the information about the thin-pool associated with the Volume Group.
+
+``` go
+
++ type ThinPoolConfig struct {
++   // Name of thin pool
++   Name string`json:"name"`
++   // OverProvisionRatio represents the ratio of overprovision that can be + allowed on thin pools
++   OverProvisionRatio float64 `json:"overprovision-ratio"`
```

Comment:
What are the possible values for overProvisionRatio? Can it be set to 0 to disable it?

Reply:
I could think the possible values as `>0` and `<100`. 
By disabling overProvisionRatio, do you mean that no overprovision limit will be set and user can provision any number of thin logical volumes on a thin-pool?

I don't think we should give an option for the user to disable it for some thin-pools.
 - Can't think of a use case why user would want to disable it for some pools. 
 -  It will add more complexity.

Reply:
possible values --> anything greater than 1.0
set to 0 to disable it --> what should be the annotation value if we set it to 0?

Reply:
I forgot to answer this question. It's reasonable to set OverProvisionRatio to equal or greater than 1.0 and regards any other values as invalid because it's not "thin" provisioning.